bokeh/models/ranges.py

```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
     this categorical range.
     """)
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```

Comment:
> I'm not sure about `start_bound` and `end_bound`. Remember, it is possible for `start > end` (i.e., if you want "reversed" range displayed) so I think `min` and `max` more clearly express the intent. Then the internals can sort out if `start` or `end` is the min, and do whatever is appropriate. Otherwise we force this on users to know and do the right thing and that seems fragile. 

I spelled it out

```python
# start < end
Range1d(start=0, end=1, start_bound=0.1, end_bound=0.9) # fine
Range1d(start=0, end=1, min=0.1, max=0.9) # fine

Range1d(start=0, end=1, start_bound=1.1, end_bound=-0.1) # error?
Range1d(start=0, end=1, min=1.1, max=-0.1) # error?

# start > end
Range1d(start=1, end=0, start_bound=1.1, end_bound=-0.1) # fine
Range1d(start=1, end=0, min=0.1, max=1.1) # fine
Range1d(start=1, end=0, max=1.1, min=0.1) # fine

Range1d(start=1, end=0, start_bound=0.1, end_bound=1.1) # error?
Range1d(start=1, end=0, min=1.1, max=0.1) # error?
```
I don't like any of these. 

New proposal:
 * Range1d
   * start - Either(Float, Datetime, Int)
   * end - Either(Float, Datetime, Int)
   * bounds - Either(Tuple(Float, Float), Tuple(Datetime, Datetime), Tuple(Int, Int), Auto, None, default=Auto)
 * DataRange1d
  * start - Float
  * end - Float
  * bounds - Either(Tuple(Float, Float), Auto, None, default=Auto)
 * FactorRange
  * factors = Either(List(String), List(Int))
  * bounds = Either(List(String), List(Int), Auto, None, default=Auto)

This gives the intuition of bounds being the same on each range so its easy to remember and lets them context switch as necessary.

Also has the advantage that if you're doing Range1d shorthand it should still be easy `Range1d(1, 0, bounds=(-0.1, 1.2))

For the Tuples, two options:
* the first value is the min and the second value is the max.
* order is irrelevant, we figure it out, larger is the max, smaller is the min.